agents/ParentshipAgent.java

```diff
@@ -0,0 +1,139 @@
+                Graql.not(
+                        Graql.var("par").isa("parentship")
+                                .rel("parentship_parent", "husband")
+                                .rel("parentship_parent", "wife")
+                ),
+                Graql.var("husband").isa("person")
+                        .has("email", Graql.var("husband-email")),
+                Graql.var("wife").isa("person")
+                        .has("email", Graql.var("wife-email")),
+                Graql.var().isa("locates")
+                        .rel("locates_located", Graql.var("m"))
+                        .rel("locates_location", Graql.var("city"))
+        ).get().sort("marriage-id");
+
+        GraknClient.Session session = context.getIterationGraknSessionFor(city.getCountry().getContinent().getName());
+
+        try (GraknClient.Transaction tx = session.transaction().write()) {
+
+            city.logQuery(childrenQuery);
+            List<String> childrenEmails = tx.execute(childrenQuery)
+                    .stream()
+                    .map(conceptMap -> conceptMap.get("email").asAttribute().value().toString())
+                    .sorted()
+                    .collect(Collectors.toList());
+
+            city.logQuery(marriageQuery);
+            List<ConceptMap> marriageAnswers = tx.execute(marriageQuery);
+
+            List<String> wifeEmails = marriageAnswers
+                    .stream()
+                    .map(a -> a.get("wife-email").asAttribute().value().toString())
+                    .sorted()
+                    .collect(toList());
+
+            List<String> husbandEmails = marriageAnswers
+                    .stream()
+                    .map(a -> a.get("husband-email").asAttribute().value().toString())
+                    .sorted()
+                    .collect(toList());
+
+            if (marriageAnswers.size() > 0 && childrenEmails.size() > 0) {
+                LinkedHashMap<Integer, List<Integer>> childrenPerMarriage = Allocation.allocateEvenlyToMap(childrenEmails.size(), marriageAnswers.size());
+
+                for (Map.Entry<Integer, List<Integer>> entry : childrenPerMarriage.entrySet()) {
+                    Integer marriageIndex = entry.getKey();
+                    List<Integer> children = entry.getValue();
+
+                    String motherEmail = wifeEmails.get(marriageIndex);
+                    String husbandEmail = husbandEmails.get(marriageIndex);
+
+                    List<String> childEmails = new ArrayList<>();
+                    for (Integer childIndex : children) {
+                        childEmails.add(childrenEmails.get(childIndex));
+                    }
+                    GraqlInsert parentshipQuery = buildParentshipInsertQuery(motherEmail, husbandEmail, childEmails);
+                    city.logQuery(parentshipQuery);
+                    tx.execute(parentshipQuery);
+                }
+                tx.commit();
+            }
+        }
+    }
```

Comment:
This method looks like it could be more readable by splitting the code into multiple functions that can be called on after the other to show a clear narrative of the method. This logic may be applicable to other agents. If so, I suggest refactoring them.

Reply:
I was under the impression that we wanted to leave the code as a simple block within agents, but I agree it makes for poor readability, I've refactored all agents accordingly.